File: frontend/services/dns-record-service.ts
import apiClient from "@/lib/axios";
import type { DNSRecord, DNSRecordCreate, DNSRecordUpdate, PaginatedResponse } from "@/types";

interface ListParams {
  page?: number;
  page_size?: number;
  search?: string;
  filter_type?: string;
  sort_by?: string;
  sort_order?: string;
}

export const dnsRecordService = {
  async list(zoneId: number, params: ListParams = {}): Promise<PaginatedResponse<DNSRecord>> {
    const { data } = await apiClient.get<PaginatedResponse<DNSRecord>>(
      `/api/hosted-zones/${zoneId}/records`,
      { params }
    );
    return data;
  },

  async get(recordId: number): Promise<DNSRecord> {
    const { data } = await apiClient.get<DNSRecord>(`/api/records/${recordId}`);
    return data;
  },

  async create(zoneId: number, payload: DNSRecordCreate): Promise<DNSRecord> {
    const { data } = await apiClient.post<DNSRecord>(
      `/api/hosted-zones/${zoneId}/records`,
      payload
    );
    return data;
  },

  async update(recordId: number, payload: DNSRecordUpdate): Promise<DNSRecord> {
    const { data } = await apiClient.put<DNSRecord>(
      `/api/records/${recordId}`,
      payload
    );
    return data;
  },

  async delete(recordId: number): Promise<void> {
    await apiClient.delete(`/api/records/${recordId}`);
  },

  async bulkDelete(zoneId: number, ids: number[]): Promise<{ deleted: number }> {
    const { data } = await apiClient.delete<{ deleted: number }>(
      `/api/hosted-zones/${zoneId}/records/bulk`,
      { data: { ids } }
    );
    return data;
  },
};

File: backend/app/services/dns_record_service.py
from __future__ import annotations

import math
from typing import List, Optional

from fastapi import HTTPException, status
from sqlalchemy.orm import Session

from app.repositories.dns_record_repository import DNSRecordRepository
from app.repositories.hosted_zone_repository import HostedZoneRepository
from app.schemas.dns_record import (
    DNSRecordCreate,
    DNSRecordListResponse,
    DNSRecordResponse,
    DNSRecordUpdate,
)


class DNSRecordService:
    def __init__(self, db: Session):
        self.record_repo = DNSRecordRepository(db)
        self.zone_repo = HostedZoneRepository(db)

    def _get_zone_or_404(self, zone_id: int):
        zone = self.zone_repo.get_by_id(zone_id)
        if not zone:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Hosted zone with id {zone_id} not found",
            )
        return zone

    def list_records(
        self,
        hosted_zone_id: int,
        page: int = 1,
        page_size: int = 25,
        search: Optional[str] = None,
        filter_type: Optional[str] = None,
        sort_by: str = "name",
        sort_order: str = "asc",
    ) -> DNSRecordListResponse:
        self._get_zone_or_404(hosted_zone_id)
        items, total = self.record_repo.list_by_zone(
            hosted_zone_id=hosted_zone_id,
            page=page,
            page_size=page_size,
            search=search,
            filter_type=filter_type,
            sort_by=sort_by,
            sort_order=sort_order,
        )
        total_pages = max(1, math.ceil(total / page_size)) if total > 0 else 1
        return DNSRecordListResponse(
            items=[DNSRecordResponse.model_validate(r) for r in items],
            total=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages,
        )

    def get_record(self, record_id: int) -> DNSRecordResponse:
        record = self.record_repo.get_by_id(record_id)
        if not record:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"DNS record with id {record_id} not found",
            )
        return DNSRecordResponse.model_validate(record)

    def create_record(self, hosted_zone_id: int, data: DNSRecordCreate) -> DNSRecordResponse:
        self._get_zone_or_404(hosted_zone_id)
        record = self.record_repo.create(hosted_zone_id, data)
        self.zone_repo.update_record_count(hosted_zone_id)
        return DNSRecordResponse.model_validate(record)

    def update_record(self, record_id: int, data: DNSRecordUpdate) -> DNSRecordResponse:
        record = self.record_repo.get_by_id(record_id)
        if not record:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"DNS record with id {record_id} not found",
            )
        record = self.record_repo.update(record, data)
        return DNSRecordResponse.model_validate(record)

    def delete_record(self, record_id: int) -> None:
        record = self.record_repo.get_by_id(record_id)
        if not record:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"DNS record with id {record_id} not found",
            )
        zone_id = record.hosted_zone_id
        self.record_repo.delete(record)
        self.zone_repo.update_record_count(zone_id)

    def bulk_delete_records(self, hosted_zone_id: int, record_ids: List[int]) -> int:
        self._get_zone_or_404(hosted_zone_id)
        deleted = self.record_repo.bulk_delete(record_ids, hosted_zone_id)
        self.zone_repo.update_record_count(hosted_zone_id)
        return deleted
